test/jb_eth_payment_terminal_store/record_payment_from.test.js

```diff
@@ -0,0 +1,537 @@
+import jbPayDelegate from '../../artifacts/contracts/mocks/JBPayDelegateMock.sol/JBPayDelegateMock.json';
+import jbPrices from '../../artifacts/contracts/JBPrices.sol/JBPrices.json';
+import jbProjects from '../../artifacts/contracts/JBProjects.sol/JBProjects.json';
+import jbTokenStore from '../../artifacts/contracts/JBTokenStore.sol/JBTokenStore.json';
+
+describe('JBETHPaymentTerminalStore::recordPaymentFrom(...)', function () {
+  const PROJECT_ID = 2;
+  const AMOUNT = ethers.FixedNumber.fromString('4398541.345');
+  const WEIGHT = ethers.FixedNumber.fromString('900000000.23411');
+  const WEIGHTED_AMOUNT = WEIGHT.mulUnsafe(AMOUNT);
+
+  async function setup() {
+    const [deployer, terminal, payer, beneficiary] = await ethers.getSigners();
+
+    const mockJbPrices = await deployMockContract(deployer, jbPrices.abi);
+    const mockJbProjects = await deployMockContract(deployer, jbProjects.abi);
+    const mockJbDirectory = await deployMockContract(deployer, jbDirectory.abi);
+    const mockJbFundingCycleStore = await deployMockContract(deployer, jBFundingCycleStore.abi);
+    const mockJbFundingCycleDataSource = await deployMockContract(
+      deployer,
+      jbFundingCycleDataSource.abi,
+    );
+    const mockJbPayDelegate = await deployMockContract(deployer, jbPayDelegate.abi);
+    const mockJbTokenStore = await deployMockContract(deployer, jbTokenStore.abi);
+    const mockJbController = await deployMockContract(deployer, jbController.abi);
+
+    const jbEthPaymentTerminalStoreFactory = await ethers.getContractFactory(
+      'JBETHPaymentTerminalStore',
+    );
+    const jbEthPaymentTerminalStore = await jbEthPaymentTerminalStoreFactory.deploy(
+      mockJbPrices.address,
+      mockJbProjects.address,
+      mockJbDirectory.address,
+      mockJbFundingCycleStore.address,
+      mockJbTokenStore.address,
+    );
+
+    const blockNum = await ethers.provider.getBlockNumber();
+    const block = await ethers.provider.getBlock(blockNum);
+    const timestamp = block.timestamp;
+
+    return {
+      terminal,
+      payer,
+      beneficiary,
+      mockJbController,
+      mockJbDirectory,
+      mockJbFundingCycleStore,
+      mockJbFundingCycleDataSource,
+      mockJbPayDelegate,
+      jbEthPaymentTerminalStore,
+      timestamp,
+    };
+  }
+
+  /* Happy path tests with terminal access */
+
+  it('Should record payment without a datasource', async function () {
+    const {
+      terminal,
+      payer,
+      beneficiary,
+      mockJbController,
+      mockJbDirectory,
+      mockJbFundingCycleStore,
+      jbEthPaymentTerminalStore,
+      timestamp,
+    } = await setup();
+
+    // Set terminal address
+    await jbEthPaymentTerminalStore.claimFor(terminal.address);
+
+    const reservedRate = 0;
+    const packedFCM = packFundingCycleMetadata({ pausePay: 0, reservedRate: reservedRate });
```

Comment:
nit) consider: metadata here and elsewhere. I'm not a fan of acronyms personally.